Toric/Mathlib/Geometry/Convex/Cone/PolyhedralCone.lean

```diff
@@ -0,0 +1,225 @@
+/-
+Copyright (c) 2025 Justus Springer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Justus Springer
+-/
+import Toric.Mathlib.Geometry.Convex.Cone.Dual
+import Toric.Mathlib.Algebra.Order.Nonneg.Module
+
+/-!
+
+# Polyhedral cones
+
+Given a bilinear pairing `p` between two `R`-modules `M` and `N`, we define
+polyhedral cones to be pointed cones in `N` that are the dual of a finite set
+in `M` (this means they are the intersection of finitely many halfspaces).
+
+The main statement is that if both `M` and `N` are finite and the pairing is injective
+in both arguments, then polyhedral cones are precisely the finitely generated cones, see
+`IsPolyhedral_iff_fg`. Moreover, we obtain that the dual of a polyhedral cone is again polyhedral
+(`IsPolyhedral_dual_of_IsPolyhedral`) and that the double dual of a polyhedral cone is the
+cone itself (`IsPolyhedral_dual_dual`).
+
+-/
+
+
+namespace PointedCone
+
+open Submodule hiding span
+
+section PartialOrder
+
+variable {R M N : Type*} [CommRing R] [PartialOrder R] [IsOrderedRing R] [AddCommGroup M]
+  [AddCommGroup N] [Module R M] [Module R N] {p : M →ₗ[R] N →ₗ[R] R}
+
+variable (p) in
+def IsPolyhedral (c : PointedCone R N) : Prop :=
+  ∃ t : Finset M, c = dual' p t
+
+lemma IsPolyhedral_def {c : PointedCone R N} :
+    IsPolyhedral p c ↔ ∃ t, t.Finite ∧ c = dual' p t :=
+  ⟨fun ⟨t, ht⟩ => ⟨t, t.finite_toSet, ht⟩,
+   fun ⟨_, ht1, ht2⟩ =>⟨ht1.toFinset, ht1.coe_toFinset.symm ▸ ht2⟩⟩
+
+lemma IsPolyhedral_dual_of_Finite {t : Set M} (h : t.Finite) :
+    IsPolyhedral p (dual' p t) := ⟨h.toFinset, by simp⟩
+
+lemma IsPolyhedral_dual_of_FG {c : PointedCone R M} (hc : c.FG) :
+    IsPolyhedral p (dual' p (c : Set M)) := by
+  obtain ⟨S, rfl⟩ := hc
+  rw [dual_span]
+  exact ⟨S, rfl⟩ 
+
+theorem IsPolyhedral_top : IsPolyhedral p (⊤ : PointedCone R N) := ⟨∅, by simp⟩
+
+@[simp]
+theorem IsPolyhedral_dual_dual {c : PointedCone R N} (hc : IsPolyhedral p c) :
+    dual' p (dual' p.flip (c : Set N)) = c := by
+  obtain ⟨t,rfl⟩ := hc
+  exact dual_dual_dual_eq_dual
+
+theorem IsPolyhedral_dual_inj {c₁ c₂ : PointedCone R N} (hc₁ : IsPolyhedral p c₁)
+    (hc₂ : IsPolyhedral p c₂) (h : dual' p.flip c₁ = dual' p.flip c₂) : c₁ = c₂ := by
+  rw [← IsPolyhedral_dual_dual hc₁, ← IsPolyhedral_dual_dual hc₂, h]
+
+end PartialOrder
+
+
+section LinearOrder
+
+variable {𝕜 M N : Type*} [Field 𝕜] [LinearOrder 𝕜] [IsStrictOrderedRing 𝕜] [AddCommGroup M]
+  [AddCommGroup N] [Module 𝕜 M] [Module 𝕜 N] {p : M →ₗ[𝕜] N →ₗ[𝕜] 𝕜}
+
+/-- If the module `M` is finite and the pairing induces an injection `N` into `M →ₗ[𝕜] 𝕜`,
+then the zero cone in `N` is polyhedral. -/
+theorem IsPolyhedral_bot [Module.Finite 𝕜 M] (hp : Function.Injective p.flip) :
```

Comment:
Do we have a use case where `Injective p.flip` holds, but `p` isn't a perfect pairing?

Reply:
I can't think of any, but wanted to use minimal assumptions here.